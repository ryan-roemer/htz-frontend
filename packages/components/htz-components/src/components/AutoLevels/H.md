### AutoLevels with Section and H

`<LevelContext />`, `<Section />` and `<H />` allow us to automatically handle Heading levels in our app

These components are based on: [Managing Heading Levels In Design Systems](https://medium.com/@Heydon/managing-heading-levels-in-design-systems-18be9a746fa3)

In order to use the components our app needs to be wrapped in LevelProvider exported from LevelContext

The `<H />` Component takes care of headings h2-h6 according to the number of sections its nested in.

The user is responsible for creating the main heading `<h1 />`>

**Basic Usage**

```jsx static
<div>
  <h1>h1 is the user's responsibility</h1>
  <H>The first level H renders is a h2</H>
  <p>Some content....</p>
  <Section>
    <H>Rendering and H inside a section will give us a h3</H>
    <p>Some content....</p>
    <Section>
      <H>Rendering and H inside a nested section will give us a h4</H>
      <p>Some content....</p>
    </Section>
  </Section>
</div>
```

```jsx
<div>
  <h1>h1 is the user's responsibility</h1>
  <H>The first level H renders is a h2</H>
  <p>Some content....</p>
  <Section>
    <H>Rendering and H inside a section will give us a h3</H>
    <p>Some content....</p>
    <Section>
      <H>Rendering and H inside a nested section will give us a h4</H>
      <p>Some content....</p>
    </Section>
  </Section>
</div>
```

**Usage With offSet**

`offSet` prop allows manipulating the heading level

```jsx
<div>
  <h1>h1 is the user's responsibility</h1>
  <H offSet={1}>I should be a h2 but I was offSet to be a h3</H>
  <p>Some content....</p>
  <Section>
    <H offSet={3}>I should be a h3 but I was offSet to be h6 </H>
    <p>Some content....</p>
    <Section>
      <H offSet={-2}>I should be a h4 but I was offSet to be h2</H>
      <p>Some content....</p>
    </Section>
  </Section>
</div>
```

**Usage with Section props**

`offSet` prop allows manipulating the heading level

```jsx
<div>
  <h1>h1 is the user's responsibility</h1>
  <H offSet={1}>I should be a h2 but I was offSet to be a h3</H>
  <p>Some content....</p>
  <Section>
    <H offSet={3}>I should be a h3 but I was offSet to be h6 </H>
    <p>Some content....</p>
    <Section>
      <H offSet={-2}>I should be a h4 but I was offSet to be h2</H>
      <p>Some content....</p>
    </Section>
  </Section>
</div>
```
