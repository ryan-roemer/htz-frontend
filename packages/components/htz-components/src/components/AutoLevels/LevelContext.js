import React from 'react';

const Level = React.createContext(1);

export const LevelProvider = Level.Provider;

export const LevelConsumer = Level.Consumer;
