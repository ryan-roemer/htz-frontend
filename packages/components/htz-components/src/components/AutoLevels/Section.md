### AutoLevels with Section and H

See detailed Docs about auto levels in `<H />` section

**Usage with props**

```jsx static
<div>
  <h1>h1</h1>
  <Section>
    <H>
      By default the Section will render a `<section />` tag
    </H>
    <p>Some content....</p>
  </Section>
  <Section tagName="div">
    <H>We can pass tagName prop and Section will render as that tag</H>
    <p>Some content....</p>
    <Section isFragment>
      <H>We can also pass isFragment prop to render as a react fragment</H>
      <p>Some content....</p>
    </Section>
  </Section>
</div>
```

```jsx
<div>
  <h1>h1</h1>
  <Section>
    <H>By default the Section will render a section tag</H>
    <p>Some content....</p>
  </Section>
  <Section tagName="div">
    <H>We can pass tagName prop and Section will render as that tag</H>
    <p>Some content....</p>
    <Section isFragment>
      <H>
        We can also pass isFragment prop to render as a react fragment,
        basically allowing us to manipulate the heading level, use with caution!
      </H>
      <p>Some content....</p>
    </Section>
  </Section>
</div>
```
