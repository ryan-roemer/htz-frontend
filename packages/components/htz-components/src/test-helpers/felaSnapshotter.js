import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
