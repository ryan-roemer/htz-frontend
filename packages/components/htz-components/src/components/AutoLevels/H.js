import React from 'react';
import PropTypes from 'prop-types';
import { LevelConsumer, } from './LevelContext';

H.propTypes = {
  /**
   * The offset from the calculated heading level.
   * e.g: the calculated level is 3 and the offset is 1, the heading level will be 4.
   * The offset can be negative.
   * The final Heading level can be between 2 and 6,
   * e.g the calculated level is 10, the heading level will be 6.
   */
  offset: PropTypes.number,
  /** Forces the generated element to be an H1 */
  isH1: PropTypes.bool,
};

H.defaultProps = {
  offset: 0,
  isH1: false,
};
function H({ offset, isH1, ...props }) {
  return (
    <LevelConsumer>
      {level => {
        const hLevel = isH1
          ? 1
          : Math.max(Math.min(Math.round(level + offset), 6), 2);
        const Heading = `h${hLevel}`;
        return <Heading {...props} />;
      }}
    </LevelConsumer>
  );
}
export default H;
