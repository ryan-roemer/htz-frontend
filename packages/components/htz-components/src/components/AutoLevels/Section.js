import React, { Fragment, } from 'react';
import PropTypes from 'prop-types';
import { LevelConsumer, LevelProvider, } from './LevelContext';

Section.propTypes = {
  /** The Sections children nodes   */
  children: PropTypes.node,
  /** Should the Section Render as a react Fragment?   */
  isFragment: PropTypes.bool,
  /** The HTML tag to render the `<Section />` as */
  tagName: PropTypes.string,
};

Section.defaultProps = {
  children: null,
  isFragment: false,
  tagName: 'section',
};

function Section({ children, isFragment, tagName, ...props }) {
  const Tag = isFragment ? Fragment : tagName;
  return (
    <LevelConsumer>
      {level => (
        <LevelProvider value={level + 1}>
          <Tag {...props}>{children}</Tag>
        </LevelProvider>
      )}
    </LevelConsumer>
  );
}

export default Section;
