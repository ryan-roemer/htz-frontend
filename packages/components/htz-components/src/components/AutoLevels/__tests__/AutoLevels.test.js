import React from 'react';
import felaSnapshotter from '../../../test-helpers/felaSnapshotter';
import { LevelProvider, } from '../LevelContext';
import H from '../H';
import Section from '../Section';

describe('<AutoLevels>', () => {
  it('renders correctly H and Sections with correct heading level in different scenarios', () => {
    const { component, styles, } = felaSnapshotter(
      <LevelProvider value={2}>
        <H>should be h2</H>
        <Section>
          <H>Should be h3</H>
          <Section tagName="div">
            <H>Should be h4</H>
            <H offSet={1}>Should be h5 because of offSet</H>
            <Section isFragment>
              <H>Should be h5 because of fragment</H>
            </Section>
          </Section>
        </Section>
      </LevelProvider>
    );
    expect(component).toMatchSnapshot();
    expect(styles).toMatchSnapshot();
  });
});
