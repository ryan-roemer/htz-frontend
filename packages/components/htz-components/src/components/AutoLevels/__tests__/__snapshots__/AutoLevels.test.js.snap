// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoLevels> renders correctly H and Sections with correct heading level in different scenarios 1`] = `
<div>
  <h2>
    should be h2
  </h2>
  <section>
    <h3>
      Should be h3
    </h3>
    <div>
      <h4>
        Should be h4
      </h4>
      <h5>
        Should be h5 because of offset
      </h5>
      <h5>
        Should be h5 because of fragment
      </h5>
    </div>
  </section>
</div>
`;

exports[`<AutoLevels> renders correctly H and Sections with correct heading level in different scenarios 2`] = `""`;
